TGUI Select Equipment menu

TheTrigg opened the pull request:
## About The Pull Request

TGUI Select Equipment menu - Or, _how my mediocre DM and JS skills will get me bullied by TG maintainers._

~I have no notion of JS coding standards, so Stylemistake will probably want to gouge his eyes out (and probably close this PR) - All I did was have ESLint not scream.~
Actually Stylemistake helped me gigantically and I wouldn't ever have brought this PR up to standards without him, thank you so much!


What this PR also does:
- ~Added a new helper proc: sortAssocList - it is identical to sortList, but guess what - it sorts an assoc list. wow.~ Ended up not needing it so I got rid of it for atomization's sake
- Slightly optimized get_flat_human_icon()
- Added a new proc to dummy code - clear_human_dummy() to clear an unused dummy key; also refactored dummy code a bit for better dummy reusing

Not the most atomic of PRs, yes, but I needed most of the changes above to make this one work. Well, actually only the last one was _really_ needed (but ugly workarounds are always an option I guess)

~NOTE: The old Select Equipment menu has NOT BEEN REMOVED YET for testmerge/comparison purposes.~ It's gone now

**Also, feedback is always appreciated**. I'm not the best with UX.

## Why It's Good For The Game

What bothered me the most about the basic menu was the lack of a preview.


https://user-images.githubusercontent.com/36010999/111222064-18b72880-85ba-11eb-8d59-17225a0329b1.mp4

Now there's one, and paired with a pretty UI and a search bar.
There's also some outfits that weren't displayed in the old menu because they had duplicate names.

About this demo video in specific: 
- Ideally the UI shouldn't close when transforming from an observer, that's a bug with the TGUI subsystem and a rabbithole I did not want to jump into
- I forgot to demo the search bar but you can guess what it does, right?


## Changelog
:cl: Trigg, stylemistake
admin: Admins just got a new TGUI Select Equipment menu
/:cl:

TheTrigg committed:
fine I'll commit

TheTrigg committed:
see I already fucked up

TheTrigg committed:
dummy now looks like the target

TheTrigg committed:
dynamic, as all things should be

TheTrigg committed:
this one can change during runtime

TheTrigg committed:
UI act, also rearranged the UI a bit

TheTrigg committed:
handling custom outfits is gonna be harder than I thought

TheTrigg committed:
custom outfits are done, sorta

TheTrigg committed:
don't need this anymore

TheTrigg committed:
now properly sorted!

TheTrigg committed:
turns out that didnt work

TheTrigg committed:
tbh, I don't think that'll be really needed

TheTrigg committed:
Is it... PR ready?
probably not, time for a public shaming session on TG's repo

TheTrigg committed:
not before this

TheTrigg committed:
woop, proper naming

TriggeredBoi commented on code/__HELPERS/_lists.dm:
```diff
@@ -331,6 +331,10 @@
 	return sortTim(L.Copy(), cmp)
 
+//sortList but with assoc lists
+/proc/sortAssocList(list/L, cmp=/proc/cmp_text_asc)
```
Didn't know whether to add an optional argument for assoc lists in the existing `sortList` or to make another proc, so I flipped a coin. It landed on tails.

stylemistake replied:
Reasonable choice.

tralezab commented:
well done, don't leave the old one

stylemistake commented:
Very impressive.

> I have no notion of JS coding standards.

I suspect you just have no intention to maintain it if you've written it that way.

tralezab commented:
Yeah uh wow that JS I just looked and that is uh... whew jesus

stylemistake commented:
I have no questions for dmcode. Server-side sorting is used sparingly, so should be fine. Usage of static data, while not canon, is justifiable for performance reasons.

I really just dislike your JS code. Please at least format it properly.

tralezab commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+
+          <Flex.Item grow={1} basis={0}>
+            <Section height="15%" mb={0}>
+              {displayTabs}
```
This applies to essentially all of your js, these need to be components instead of variables, see OperatingComputer.js for a good example (lines 44, 47)


do this for all offenders in this function!

stylemistake commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,164 @@
+			if(!istype(O))
+				return
+			selected_outfit = O.type //the typepath - not the object
+			update_static_data(user.mob)
```
Suggestion:

Here, set `. = TRUE` instead of this call, to signal that ui_data should be sent.

Rename ui_static_data proc to ui_data.

In ui_interact, after the line which opens the ui, `ui.set_autoupdate(FALSE)`

tralezab commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+    );
+  };
+
+  const outfitCategories = Object.keys(outfits);
```
You can send data better so that you don't need to do this but honestly not the end of the world

tralezab commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,164 @@
+		cached_outfits = list()
+		cached_outfits["General"] = list(/datum/outfit = "Naked") + make_outfit_entries(subtypesof(/datum/outfit) - typesof(/datum/outfit/job) - typesof(/datum/outfit/plasmaman))
+		cached_outfits["Jobs"] = make_outfit_entries(typesof(/datum/outfit/job))
+		cached_outfits["Plasmamen Outfits"] = make_outfit_entries(typesof(/datum/outfit/plasmaman))
```
These need to be nested inside a list so they are sent to the data as a mappable list and not an object you need to get the keys for and do annoying access with

TriggeredBoi replied:
Not sure if you're talking about the categories or the outfit entries so I'll address both:

I think it's better for `cached_outfits` to be sent as an object as I can actually use the object methods as both keys and names for the Tabs as seen [here](https://github.com/tgstation/tgstation/pull/57673/files#diff-0931ef83d6077b6246c7d86ee4650436bea50aa7397bc02f5ba86145d5d9f49aR56-R57) - Unless doing so is cursed instead of clever (like I thought it was when I did it)

As for the outfit entries, the main reason I made them assoc lists in the first place is because sorting them in a list of lists would be both hard and ugly. Though I've still gotta find a way to implement favorites so I'll probably end up needing to implement custom sorting either way. ~Brain hurty please help~

stylemistake replied:
oh my god, what are you doing

tralezab commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+
+
+  // outfit selection
+  const selectOutfit = outfitPath => {
```
i think this just obfuscates things

TriggeredBoi replied:
How should I call both `setSelectedOutfit` and `act` then?

tralezab commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+    );
+
+
+  const outfitButton = outfit => {
```
needs to be moved off into its own function

stylemistake replied:
No, it just needs to go back to where it belongs - in `.map()` call.

tralezab commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+    );
+  };
+
+  const displayedOutfits
```
needs to be moved off into its own function/removed since i think it just obfuscates things. honestly this probably shows the most issues with the js: some things oddly too specific, and what things do being hidden in variables in something that is hidden in a variable. basically just unreadable in its entirety

tralezab reviewed (changes requested):
You need to go run bin/tgui --fix and maybe --lint commands, read the README in tgui folder to learn how

as a side note I'm not sure if style would agree but maybe it's best if you just sticked to stack OR flex since they are in a way the same thing presented in ways to give more or less control at the cost of easy use

TheTrigg committed:
review 1 out of many
nothing unexpected to be fair

TriggeredBoi commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,149 @@
+  const searchFilter = createSearch(searchText, entry =>
+    (entry[0] + entry[1])
+  );
+  const searchBar
```
Like this? (Or am I just dense?)
```suggestion
  const searchBar = (props, context) => {
    const { act, data } = useBackend(context);
    return (
      <Input
        fluid
        autoFocus
        placeholder="Search"
        value={searchText}
        onInput={(e, value) => setSearchText(value)}
        mb={1} />);
  };
```

TheTrigg committed:
so if I'm not dumb, this is what I'm supposed to do

Qustinnus commented:
feature request: add a favorite panel that saves to prefs for testing maybe?

MrDoomBringer commented:
Looks really good, save for the previously-mentioned poor coding standards that you mentioned in the OP. Please try to fix that

Also, do you think we could scale down the preview by like 75%? The menu doesn't need to take up this much space.

optimumtact commented:
>I have no notion of JS coding standards

that's no problem, we can beat them into you

TheTrigg committed:
function-izing
except for searchBar because apparently it causes a memory leak

TriggeredBoi commented:
> I suspect you just have no intention of maintaining it if you've written it that way.
 
I'll do my best at maintaining it, though as you've seen I'm not good at that


> add a favorite panel that saves to prefs for testing maybe?

Will do


> Scale down the preview to 25% of its current size

At the moment the UI acts funky when any smaller than this due to some elements just refusing to scale down (especially the Scrollable section) but I'll make sure it can be smaller


> that's no problem, we can beat them into you

Fair enough

TheTrigg committed:
reworked to use lists instead of objects in JS side
kinda looks like a downgrade but hopefully I can implement favorites off of this

AnturK commented:
Custom outfits should just be saved in local storage with this, with optional export.

TheTrigg committed:
Revert "reworked to use lists instead of objects in JS side"

This reverts commit 1336b9f692b8c3c8e9eeb37d1515fcf6c6b06ff7.

TheTrigg committed:
entries are now objects

TheTrigg committed:
now well sent and well sorted
btw stylemistake did all the big brain thinking for me

TheTrigg committed:
stylemistake said he'd give me good boy points if I did this

TheTrigg committed:
the great engoodening
stylemistake spent like 4 hours of his life guiding me around this

TheTrigg committed:
minor tidbits

TheTrigg committed:
fixed custom outfits, bit of cleanup

TheTrigg committed:
GLOB.custom_outfits allows duplicate names

stylemistake commented:
Good boy points dispensed.

TriggeredBoi commented:
Ok, all that's left for me to do is to implement favorites (will do that tomorrow) and then remove the old menu; so I'm marking this PR as Ready for review (again)

\***Huge*** thanks to @stylemistake for bothering and tolerating my smoothbrain for hours and guiding me toward making the JS code decent.

I shall cherish the good boy points religiously.

Mothblocks commented on code/__HELPERS/_lists.dm:
```diff
@@ -331,6 +331,10 @@
 	return sortTim(L.Copy(), cmp)
 
+//sortList but with assoc lists
+/proc/sortAssocList(list/L, cmp=/proc/cmp_text_asc)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

TriggeredBoi replied:
Should I change the other sort procs too for consistency?

Mothblocks replied:
Just keep it here for now, atomization and all.

TriggeredBoi replied:
To be fair I'm not even using it anymore actually

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+/client/proc/cmd_select_equipment(mob/M in GLOB.mob_list)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	set name = "Select equipment but better"
+
+
+	var/datum/select_equipment/ui  = new(usr, M)
```
```suggestion
	var/datum/select_equipment/ui = new(usr, M)
```

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	var/datum/select_equipment/ui  = new(usr, M)
+	ui.ui_interact(usr)
+
+/datum/select_equipment
```
Document this.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	//used to keep track of which outfit the UI has selected
+	var/selected_name = "/datum/outfit"
+
+/datum/select_equipment/New(usr, mob/M)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+/datum/select_equipment/New(usr, mob/M)
+	user = CLIENT_FROM_VAR(usr)
+
+	if(!(ishuman(M) || isobserver(M)))
```
```suggestion
	if(!ishuman(M) || !isobserver(M))
```

stylemistake replied:
bad logic @Mothblocks

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+/datum/select_equipment/proc/init_dummy()
+	dummy_key = "selectequipmentUI_[target]"
+	dummy = generate_or_wait_for_human_dummy(dummy_key)
+	var/mob/living/carbon/C = target
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+#define OUTFIT_ENTRY(category, path, name, priority) list("category" = category, "path" = path, "name" = name, "priority" = priority)
+//passes a ref instead of path, every entry needs some sort of unique identifier
+//and GLOB.custom_outfits allows duplicate names
+#define OUTFIT_CUSTOM_ENTRY(category, ref, name, priority) list("category" = category, "ref" = ref, "name" = name, "priority" = priority)
```
Undef this.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+
+/datum/select_equipment/proc/make_outfit_entries(category="General", list/L)
+	var/list/entries = list()
+	for(var/path in L)
+		var/datum/outfit/O = path
```
Use as anything. Also, better variable names.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	for(var/path in L)
+		var/datum/outfit/O = path
+		var/priority = 0
+		if(path == /datum/outfit/job/roboticist) //dummy check here until I get favorites working
```
Then your PR isn't ready, make it a draft.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+
+
+//GLOB.custom_outfits lists outfit *objects* so we'll need to do some custom handling for it
+/datum/select_equipment/proc/make_custom_outfit_entries(list/L)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+//GLOB.custom_outfits lists outfit *objects* so we'll need to do some custom handling for it
+/datum/select_equipment/proc/make_custom_outfit_entries(list/L)
+	var/list/entries = list()
+	for(var/datum/outfit/O in L)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	var/list/data = list()
+	if(!cached_outfits)
+		cached_outfits = list()
+		cached_outfits += list(OUTFIT_ENTRY("General", /datum/outfit, "Naked", 2))
```
Magic 2. Move it to a define.

Mothblocks replied:
Ideally these magic values would be sent through to ui_static_data as well.

TriggeredBoi replied:
No need to send them to the UI, they're just there [as priorities for sorting](https://github.com/tgstation/tgstation/pull/57673/files#diff-0931ef83d6077b6246c7d86ee4650436bea50aa7397bc02f5ba86145d5d9f49aR100)

TriggeredBoi replied:
Nevermind; after I finish favorites, `priorities` will turn into a bool and the magic numbers will be gone

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	. = TRUE
+	switch(action)
+		if("preview")
+			var/datum/outfit/O = resolve_outfit(params["path"])
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+			selected_outfit = O
+
+		if("applyoutfit")
+			var/datum/outfit/O = resolve_outfit(params["path"])
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+			user.outfit_manager()
+
+
+/client/proc/admin_apply_outfit(mob/M, dresscode)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+
+
+/client/proc/admin_apply_outfit(mob/M, dresscode)
+	if(!(ishuman(M) || isobserver(M)))
```
```suggestion
	if(!ishuman(M) || !isobserver(M))
```

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+		return
+
+	var/delete_pocket
+	var/mob/living/carbon/human/H
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+	var/delete_pocket
+	var/mob/living/carbon/human/H
+	if(isobserver(M))
+		H = M.change_mob_type(/mob/living/carbon/human, null, null, TRUE)
```
Use named arguments and drop the `null, null`.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,197 @@
+				delete_pocket = TRUE
+
+	SSblackbox.record_feedback("tally", "admin_verb", 1, "Select Equipment") //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
+	for (var/obj/item/I in H.get_equipped_items(delete_pocket))
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/mob/living/carbon/human/dummy.dm:
```diff
@@ -40,13 +40,26 @@ GLOBAL_LIST_EMPTY(dummy_mob_list)
+	if(!slotkey)
+		return
+
+	var/mob/living/carbon/human/dummy/D = GLOB.human_dummy_list[slotkey]
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on tgui/packages/tgui/interfaces/SelectEquipment.js:
```diff
@@ -0,0 +1,173 @@
+import { useBackend, useLocalState } from '../backend';
+import { createSearch } from 'common/string';
+import { Box, Button, Tabs, Section, Input, Stack } from '../components';
```
```suggestion
import { Box, Button, Input, Section, Stack, Tabs } from '../components';
```

Alphabetical order.

TheTrigg committed:
trigg sucks at descriptive names
tho in my defense admin_apply_outfit was straight copypaste from the old menu

TheTrigg committed:
unified outfit_entry, defined magic numbers,
documentation, reverted bad logic, UI now opens again

TheTrigg committed:
whoops

TheTrigg committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into trigg-attempts-to-do-tgui-once-more

TheTrigg committed:
favorites!

TheTrigg committed:
last naming problems I think
also yes I got rid of sortAssocList

TriggeredBoi commented:
Ok favorites are implemented and I'm pretty sure all other issues are sorted so uh... @Mothblocks re-review pls?

stylemistake commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -77,36 +77,30 @@
-		return list("category" = category, "ref" = identifier, "name" = name, "priority" = priority)
-	return list("category" = category, "path" = identifier, "name" = name, "priority" = priority)
+/datum/select_equipment/proc/outfit_entry(category, identifier, name, priority=FALSE, custom_entry=FALSE)
+	return list("category" = category, custom_entry?"ref":"path" = identifier, "name" = name, "priority" = priority)
```
```
list(custom_entry?"ref":"path" = identifier)
```

It work like that? Absolutely cursed.

Mothblocks replied:
depressingly, yes. but this is unreadable, move it out to an if.

Mothblocks commented:
> re-review pls?

Undraft if ready 🙂

TheTrigg committed:
pretty icons, pretty code

TheTrigg committed:
OutfitDisplay component, started moving the favorite icon
literally only committing so I can reset after I fuck up this next part

TheTrigg committed:
hey I did it

TheTrigg committed:
moved the UI around

stylemistake committed:
Update SelectEquipment.js

stylemistake committed:
Update SelectEquipment.js

TriggeredBoi commented:
This is probably the final look of the UI, marking this as ready for review for hopefully the last time

TheTrigg committed:
farewell old menu

TheTrigg committed:
am dumb

TheTrigg committed:
surprisingly, this one wasn't me

stylemistake commented:
Note for other mains: robust_dress_shop is used in at least 3 more places

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,230 @@
+/datum/select_equipment/proc/make_outfit_entries(category="General", list/outfit_list)
+	var/list/entries = list()
+	for(var/path as anything in outfit_list)
+		var/datum/outfit/O = path
```
Better variable name.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,230 @@
+//GLOB.custom_outfits lists outfit *objects* so we'll need to do some custom handling for it
+/datum/select_equipment/proc/make_custom_outfit_entries(list/outfit_list)
+	var/list/entries = list()
+	for(var/datum/outfit/O as anything in outfit_list)
```
Better variable name.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,230 @@
+	var/list/data = list()
+	if(!cached_outfits)
+		cached_outfits = list()
+		cached_outfits += list(outfit_entry("General", /datum/outfit, "Naked", TRUE))
```
Named argument for the TRUE, so it's obvious what it is referring to.

Mothblocks commented on code/modules/admin/verbs/selectequipment.dm:
```diff
@@ -0,0 +1,230 @@
+			user.prefs.save_preferences()
+
+/client/proc/admin_apply_outfit(mob/target, dresscode)
+	if(!(ishuman(target) || isobserver(target)))
```
```suggestion
	if(!ishuman(target) || !isobserver(target))
```

TriggeredBoi replied:
Then that check will never work

Mothblocks replied:
This is exactly the same as the check you have now.

Mothblocks replied:
Oh wait, my mistake. I updated the suggestion.

Mothblocks reviewed (changes requested):
Last nitpicks.

TheTrigg committed:
probably the last commit

TriggeredBoi commented:
Done :D
